applications: machine_learning: Add information about power management (#6346)
nrfconnect/sdk-nrf: applications/machine_learning/README.rst

The application displays LED effects that correspond to the machine learning results.
For more detailed information, see the `User interface`_ section.

Power management
================

Reducing power consumption is important for every battery-powered device.

In the nRF Machine Learning application, you can suspend or turn off modules if they are not in use.

> The application uses :ref:`caf_power_manager` module for this purpose, which means that Zephyr power management is forced to :c:enumerator:`PM_STATE_ACTIVE` when device is in either the active or the suspended state, but the power off state is forced directly by :ref:`caf_power_manager` as Zephyr's :c:enumerator:`PM_STATE_SOFT_OFF`.

Reviewer: ```suggestion
The application uses :ref:`caf_power_manager` module for this purpose.
This means that Zephyr power management is forced to the :c:enumerator:`PM_STATE_ACTIVE` state when the device is in either the active or the suspended state, but the power off state is forced directly by :ref:`caf_power_manager` as Zephyr's :c:enumerator:`PM_STATE_SOFT_OFF`.
```

> The application uses :ref:`caf_power_manager` module for this purpose, which means that Zephyr power management is forced to :c:enumerator:`PM_STATE_ACTIVE` when device is in either the active or the suspended state, but the power off state is forced directly by :ref:`caf_power_manager` as Zephyr's :c:enumerator:`PM_STATE_SOFT_OFF`.

Reviewer 2: ```suggestion
The application uses :ref:`caf_power_manager` module for this purpose.
This means that Zephyr power management is forced to the :c:enumerator:`PM_STATE_ACTIVE` state when the device is either in the active or suspended state, but the power off state is forced directly by :ref:`caf_power_manager` as Zephyr's :c:enumerator:`PM_STATE_SOFT_OFF` state.
```

> * In the :c:enumerator:`PM_STATE_ACTIVE` state the device is in working condition, Bluetooth is advertising if required, all the connections are maintained.

Reviewer: ```suggestion
* In the :c:enumerator:`PM_STATE_ACTIVE` state, the device is in working condition, Bluetooth is advertising if required, and all the connections are maintained.
```

> * In the :c:enumerator:`PM_STATE_ACTIVE` state the device is in working condition, Bluetooth is advertising if required, all the connections are maintained.

Reviewer 2: ```suggestion
* In the :c:enumerator:`PM_STATE_ACTIVE` state, the device is in working condition, Bluetooth is advertising whenever required and all the connections are maintained.
```

* In the :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, the device maintains the active Bluetooth connection.
* In the :c:enumerator:`POWER_MANAGER_LEVEL_OFF` state, the CPU is switched to the off mode.

> In both cases, most of the functionalities are disabled.

Reviewer: You say `both`, but there are 3 options...
```suggestion
In all cases, most of the functionalities are disabled.
```

For example, LEDs and sensors are turned off and Bluetooth advertising is stopped.

Any button press can wake up the device.

For the Thingy:53, the sensor supports a trigger that can be used for active power management.
As long as the device detects acceleration, the board is kept in the active state.

> When the board is in :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, it can be woken up by acceleration threshold, that is by moving the device.

Reviewer: ```suggestion
When the board is in the :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, it can be woken up by acceleration threshold, that is by moving the device.
```

> When the board is in :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, it can be woken up by acceleration threshold, that is by moving the device.

Reviewer 2: ```suggestion
When the board is in the :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, it can be woken up by acceleration threshold by moving the device.
```

> You can define the amount of time after which the peripherals are suspended or powered off in :kconfig:`CONFIG_CAF_POWER_MANAGER_TIMEOUT`.

Reviewer 2: ```suggestion
You can define the time interval after which the peripherals are suspended or powered off in the :kconfig:`CONFIG_CAF_POWER_MANAGER_TIMEOUT` option.
```

The application displays LED effects that correspond to the machine learning results.
For more detailed information, see the `User interface`_ section.

Power management
================

Reducing power consumption is important for every battery-powered device.

> In the nRF Machine Learning application, you can suspend or turn off modules if they are not in use.

Reviewer 3: That's partially true - we only set a global power state for whole application, we do not have per-module power state.

Reviewer 3: Maybe:
`In the nRF Machine Learning application, application modules are automatically suspended or turned off if the device is not in use for predefined period of time.`?

> In the Machine Learning application specified modules can be suspended or turned off if they are not in use.

Author: removed

The application uses :ref:`caf_power_manager` module for that purpose.

> It means that Zephyr power management is used when device is in active on suspended state,

Reviewer 3: `is not used` :)

Author: From my perspective it is - in that state it is OS that decides if we should put the CPU down.

Reviewer 3: During module initialization we use `pm_power_state_force((struct pm_state_info){PM_STATE_ACTIVE, 0, 0});`. In that case when device is active/suspended, the Zephyr power management is actually disabled and not used since we force the active state. We change the forced state only on system off.

Reviewer 3: btw. `active or suspended state` ?

Author: ```
The application uses :ref:`caf_power_manager` module for this purpose, which means that Zephyr power management is forced to :c:enumerator:`PM_STATE_ACTIVE` when device is in either the active or the suspended state, but the power off state is forced directly by :ref:`caf_power_manager` as Zephyr's :c:enumerator:`PM_STATE_SOFT_OFF`.
```

> but power off state is forced directly by :ref:`caf_power_manager`.

Author: ok

> but power off state is forced directly by :ref:`caf_power_manager` as Zephyr's PM_STATE_SOFT_OFF.

Reviewer 3: :c:enumerator:`PM_STATE_SOFT_OFF`

> * In the :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, the device maintains the active Bluetooth connection.

Author: ```
* In the :c:enumerator:`PM_STATE_ACTIVE` state the device is in working condition, Bluetooth is advertising if required, all the connections are maintainted.
```

> * In the powered off state, the CPU is switched to the off mode.

Reviewer 4: ```suggestion
* In the :c:enumerator:`POWER_MANAGER_LEVEL_OFF` state, the CPU is switched to the off mode.
```

In both cases, most of the functionalities are disabled.

> For example, LEDs and sensors are turned off and advertising is stopped.

Reviewer 3: Thingy:53's sensor is not turned off actually (it is waiting for trigger). Maybe we should not mention sensors here?

Author: I will change it, but please note that this one is direct copy from nrf_desktop.

Any button press can wake up the device.

> For the Thingy:53 sensor supports trigger that can be used for active power management in CAF sensor sample.

Reviewer 4: ``` suggestion
For the Thingy:53 sensor supports trigger that can be used for active power management.
```
I don't understand this sentence, i will cut the ending.

As long as device acceleration is detected the board is kept in active state.

> When the board goes into sleep it can also be woken up by acceleration threshold - it means by moving the device.

Reviewer 4: ``` suggestion
When the board is in :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state it can be woken up by acceleration threshold - it means by moving the device.
```

> For the Thingy:53 sensor supports trigger that can be used for active power management.
> As long as device acceleration is detected the board is kept in active state.
> When the board is in :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state it can be woken up by acceleration threshold - it means by moving the device.

Reviewer: ```suggestion
For the Thingy:53, the sensor supports a trigger that can be used for active power management.
As long as the device detects acceleration, the board is kept in the active state.
When the board is in :c:enumerator:`POWER_MANAGER_LEVEL_SUSPENDED` state, it can be woken up by acceleration threshold, that is by moving the device.
```

    * LED blinks with an average frequency if it is connected, but is not actively forwarding data.
    * LED blinks rapidly if it is connected and is actively forwarding data.

> Power management

Author: This placement was copied from nrf_desktop layout. I may move it, but we should rethink the nrf_desktop README layout.

> Power management

Author: Moved before Firmware architecture

================

Reducing power consumption is important for every battery-powered device.

> In the Machine Learning application specified modules can be suspended or power off there are not in use.

Author: Done

> In the Machine Learning application specified modules can be suspended or power off there are not in use.

Author: Ok

> * In the suspended state, the device maintains the active connection.

Reviewer 3: `the active Bluetooth connection`?
To distinguish from data forwarding over EI data forwarder protocol (not every configuration supports Bluetooth in case of the ML application)

Author: done

> * In the suspended state, the device maintains the active connection.

Author: ok

* In the powered off state, the CPU is switched to the off mode.

In both cases, most of the functionalities are disabled.
For example, LEDs and sensors are turned off and advertising is stopped.

Any button press can wake up the device.

> For the Thingy:53 sensors supports active power management.

Reviewer 3: `sensor`? We only use accelerometer.

Reviewer 3: Supports trigger that can be used for active power management in CAF sensor sampler (to distinguish from Zephyr's power management API calls that are not supported)

Author: ok

> For the Thingy:53 sensors supports active power management.

Reviewer 4: It would be valuable to add a link to active power management from sensor manager here. I could add a label for it here https://github.com/nrfconnect/sdk-nrf/pull/6329/commits/b30dcbb1a6f5fc7de3c1d0f2c010b2e97c899702#diff-1e404cbcd69774fbac51b9195758eabdfb04431d39c7b2a3b9822ef145270f1aR304

Reviewer 4: Discussed offline. Let's skip this for now.

Author: Agreed offline to ignore this proposal.

As long as device acceleration is detected the board is kept in active state.
When the board goes into sleep it can also be woken up by acceleration threshold - it means by moving the device.

> You can define the amount of time after which the peripherals are suspended or powered off in :kconfig:`CONFIG_CAF_POWER_MANAGER_TIMEOUT`.

Reviewer 3: Actually Thingy:53 sets this to 10 seconds.

Author: Well.... it does not change the defaults. Moreover, after #6272 this settings would be reverted to default.

Reviewer 3: We can leave it as it is if this will be reverted to 120 seconds.
